tests/setup.cpp

```diff
@@ -0,0 +1,105 @@
+/*
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ *  Copyright (C) 2020-2021  The DOSBox Staging Team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "setup.h"
+
+#include <gtest/gtest.h>
+
+#include <string>
+
+// stubs
+void GFX_ShowMsg(char const *, ...) {}
+Config *control = nullptr;
+
+namespace {
+
+const parse_environ_result_t expected_empty{};
+
+TEST(ParseEnv, NoRelevantEnvVariables)
+{
+	const char *test_environ[] = {"FOO=foo", "BAR=bar", "BAZ=baz", nullptr};
+
+	EXPECT_EQ(expected_empty, parse_environ(test_environ));
+}
+
+TEST(ParseEnv, SingleValueInEnv)
+{
+	const char *test_environ[] = {
+		"SOME_NAME=value",
+		"DOSBOX_SECTIONNAME_PROPNAME=value",
+		"OTHER_NAME=other_value",
+		nullptr
+	};
+	parse_environ_result_t expected{{"SECTIONNAME", "PROPNAME=value"}};
+
+	EXPECT_EQ(expected, parse_environ(test_environ));
+}
+
+TEST(ParseEnv, PropertyOrValueCanHaveUnderscores)
+{
+	const char *test_environ[] = {
+		"DOSBOX_sec_prop_name=value",
+		"DOSBOX_sec_prop=value_name",
+		nullptr
+	};
+	parse_environ_result_t expected{{"sec", "prop_name=value"},
+	                                {"sec", "prop=value_name"}};
+
+	EXPECT_EQ(expected, parse_environ(test_environ));
+}
+
+TEST(ParseEnv, SelectVarsBasedOnPrefix)
+{
+	const char *test_environ[] = {
+		"DOSBOX_sec_prop1=value1",
+		"dosbox_sec_prop2=value2",
+		"DOSBox_sec_prop3=value3",
+		"dOsBoX_sec_prop4=value4",
+		"non_dosbox_sec_prop=val",
+		nullptr
+	};
+
+	EXPECT_EQ(4, parse_environ(test_environ).size());
+}
+
+TEST(ParseEnv, FilterOutEmptySection)
+{
+	const char *test_environ[] = {
+		"DOSBOX=value",
+		"DOSBOX_=value",
+		"DOSBOX__prop=value",
+		nullptr
+	};
+
+	EXPECT_EQ(expected_empty, parse_environ(test_environ));
+}
+
+TEST(ParseEnv, FilterOutEmptyPropname)
+{
+	const char *test_environ[] = {
+		"DOSBOX_sec=value",
+		"DOSBOX_sec_=value",
+		nullptr
+	};
+
+	EXPECT_EQ(expected_empty, parse_environ(test_environ));
+}
```

Comment:
Unit test to exercise empty or non-alpha prop value: `if (prop_set_val.empty() || !isalpha(prop_set_val[0]))`

Reply:
There is one already: `FilterOutEmptyPropname` :)

Reply:
> Are there some other cases we should include in this test

I think the empty value and non-alpha value cases. 

Dosbox_section_prop=<empty or garbage>

Reply:
But a configuration setting can contain arbitrary string in there… Empty value is perfectly valid - it's the default e.g. `midiconfig` and `language`.

Reply:
Ok, fresh morning!

I was looking at this chunk of code:

``` c++
const auto prop_set_val = rest.substr(section_delimiter + 1);
if (prop_set_val.empty() || !isalpha(prop_set_val[0]))
	continue;
```

And now understand that "val" is both the property name and its value together,  chopped off after the section underscore.

So it catches scenarios:

```
DOSBOX_SECTION_
DOSBOX_SECTION_<non alpha junk>
DOSBOX_SECTION_=
DOSBOX_SECTION_=something
```

Which indeed you have unit tests to catch 😌
Thanks for the patience, @dreamer.

I had associated `prop_set_val` to `VALUE` in:
DOSBOX_SECTIONNAME_PROPNAME=VALUE 

I suppose if there's any suggestion to help future readers, it would be to to name `prop_set_val` to `key_and_value`, or something that uses "and" or "combined" for more obvious effect. 

(I now understand you carefully intended `set` to mean the equal sign in the string.. but set is such an overloaded word it didn't register as such with me until getting to this point :-).

Reply:
Approving, and you make the call if variable name (or just a helpful comment) on this little bit. 👍

Reply:
Renaming to `propname_and_value` is a good idea, I'll implement it :)